File: lib/newsService.ts
import { NewsStory, ImpactLevel } from "./types";

const FINNHUB_API_KEY = process.env.NEXT_PUBLIC_FINNHUB_API_KEY || "sandbox_c8r2r2aad3i9v2bt8vbg";

export async function fetchMarketNews(symbols: string[] = []): Promise<NewsStory[]> {
  try {
    let allNews: any[] = [];

    if (symbols.length > 0) {
      // Fetch news for selected symbols if provided
      const today = new Date().toISOString().split('T')[0];
      const lastWeek = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];

      // We'll fetch for the first few symbols to stay within rate limits for demo
      const fetchPromises = symbols.slice(0, 3).map(symbol =>
        fetch(`https://finnhub.io/api/v1/company-news?symbol=${symbol}&from=${lastWeek}&to=${today}&token=${FINNHUB_API_KEY}`)
          .then(res => res.json())
      );

      const results = await Promise.all(fetchPromises);
      allNews = results.flat();
    }

    // If no symbol-specific news or none found, fallback to general news
    if (allNews.length === 0) {
      const response = await fetch(
        `https://finnhub.io/api/v1/news?category=general&token=${FINNHUB_API_KEY}`
      );
      if (response.ok) {
        allNews = await response.json();
      }
    }

    // Sort by time descending
    allNews.sort((a, b) => b.datetime - a.datetime);

    // Transform Finnhub data to our NewsStory format
    return allNews.slice(0, 10).map((item: any, index: number) => ({
      id: `live-${index}-${item.id}`,
      title: item.headline,
      summary: item.summary || "Stay updated with the latest market moves and financial trends.",
      imageUrl: item.image || `https://images.unsplash.com/photo-1611974717483-36aa3921500d?auto=format&fit=crop&q=80&w=800`,
      source: item.source || "Market Feed",
      timestamp: new Date(item.datetime * 1000).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
      impact: {
        level: determineImpact(item.headline),
        description: "Live market update based on current headlines.",
        affectedSymbols: extractSymbols(item.headline),
      },
      details: [
        "Real-time coverage from global news desks.",
        "Market sentiment analysis pending further data.",
        "Historical context available in the full report."
      ]
    }));
  } catch (error) {
    console.error("Error fetching live news:", error);
    return [];
  }
}

function determineImpact(headline: string): ImpactLevel {
  const lower = headline.toLowerCase();
  if (lower.includes("surge") || lower.includes("rise") || lower.includes("gain") || lower.includes("growth")) return "positive";
  if (lower.includes("fall") || lower.includes("drop") || lower.includes("slump") || lower.includes("loss")) return "negative";
  return "neutral";
}

function extractSymbols(headline: string): string[] {
  // Simple heuristic for demo purposes
  const symbols = ["AAPL", "TSLA", "NVDA", "MSFT", "AMZN", "GOOGL"];
  return symbols.filter(s => headline.includes(s));
}

File: lib/services/news_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';

class NewsService {
  static const String _finnhubApiKey = "sandbox_c8r2r2aad3i9v2bt8vbg";

  Future<List<NewsStory>> fetchMarketNews({List<String> symbols = const []}) async {
    try {
      List<dynamic> allNews = [];

      if (symbols.isNotEmpty) {
        final today = DateTime.now().toIso8601String().split('T')[0];
        final lastWeek = DateTime.now().subtract(const Duration(days: 7)).toIso8601String().split('T')[0];

        final fetchPromises = symbols.take(3).map((symbol) async {
          final url = Uri.parse('https://finnhub.io/api/v1/company-news?symbol=$symbol&from=$lastWeek&to=$today&token=$_finnhubApiKey');
          final response = await http.get(url);
          if (response.statusCode == 200) {
            return json.decode(response.body) as List<dynamic>;
          }
          return [];
        });

        final results = await Future.wait(fetchPromises);
        allNews = results.expand((x) => x).toList();
      }

      if (allNews.isEmpty) {
        final url = Uri.parse('https://finnhub.io/api/v1/news?category=general&token=$_finnhubApiKey');
        final response = await http.get(url);
        if (response.statusCode == 200) {
          allNews = json.decode(response.body) as List<dynamic>;
        }
      }

      allNews.sort((a, b) => (b['datetime'] as int).compareTo(a['datetime'] as int));

      return allNews.take(10).indexed.map((entry) {
        final index = entry.$1;
        final item = entry.$2;

        return NewsStory(
          id: 'live-$index-${item['id']}',
          title: item['headline'] ?? '',
          summary: item['summary']?.isNotEmpty == true ? item['summary'] : "Stay updated with the latest market moves and financial trends.",
          imageUrl: (item['image']?.isNotEmpty == true) ? item['image'] : "https://images.unsplash.com/photo-1611974717483-36aa3921500d?auto=format&fit=crop&q=80&w=800",
          source: item['source'] ?? "Market Feed",
          timestamp: DateTime.fromMillisecondsSinceEpoch((item['datetime'] as int) * 1000).toLocal().toString().split(' ')[1].substring(0, 5),
          impact: NewsImpact(
            level: _determineImpact(item['headline'] ?? ''),
            description: "Live market update based on current headlines.",
            affectedSymbols: _extractSymbols(item['headline'] ?? ''),
          ),
          details: _generateKeyTakeaways(item['headline'] ?? '', item['summary'] ?? ''),
        );
      }).toList();
    } catch (e) {
      print("Error fetching news: $e");
      return [];
    }
  }

  ImpactLevel _determineImpact(String headline) {
    final lower = headline.toLowerCase();
    if (lower.contains("surge") || lower.contains("rise") || lower.contains("gain") || lower.contains("growth")) return ImpactLevel.positive;
    if (lower.contains("fall") || lower.contains("drop") || lower.contains("slump") || lower.contains("loss")) return ImpactLevel.negative;
    return ImpactLevel.neutral;
  }

  List<String> _extractSymbols(String headline) {
    const symbols = ["AAPL", "TSLA", "NVDA", "MSFT", "AMZN", "GOOGL"];
    return symbols.where((s) => headline.contains(s)).toList();
  }

  List<String> _generateKeyTakeaways(String headline, String summary) {
    List<String> takeaways = [];
    if (headline.toLowerCase().contains("surge") || headline.toLowerCase().contains("rise")) {
      takeaways.add("Bullish momentum detected in recent trading sessions.");
    } else if (headline.toLowerCase().contains("fall") || headline.toLowerCase().contains("drop")) {
      takeaways.add("Bearish pressure mounting as investors react to new data.");
    } else {
      takeaways.add("Market remains cautious amid ongoing economic uncertainty.");
    }
    if (summary.isNotEmpty && summary.length > 20) {
      takeaways.add(summary.split('.').first + ".");
    } else {
      takeaways.add("Analysts are closely monitoring price action for breakout signals.");
    }
    takeaways.add("Historical volatility suggests potential for upcoming price discovery.");
    return takeaways;
  }
}
